Libraries/Text/Text/RCTTextRenderer.m

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#import "RCTTextRenderer.h"
+
+#import "RCTTextAttributes.h"
+
+@implementation RCTTextRenderer
+{
+  __weak CALayer *_layer;
+  NSTextStorage *_Nullable _textStorage;
+  CGRect _contentFrame;
+  CAShapeLayer *_highlightLayer;
+}
+
+- (instancetype)initWithLayer:(CALayer *)layer
+{
+  self = [super init];
+  if (self) {
+    _layer = layer;
+  }
+  return self;
+}
+
+- (void)setTextStorage:(NSTextStorage *)textStorage
+          contentFrame:(CGRect)contentFrame
+{
+  _textStorage = textStorage;
+  _contentFrame = contentFrame;
+}
+
+- (void)drawLayer:(CALayer *)layer
```

Comment:
I am worried about two things:
 * How can we deduplicate the code in this method and the similar method in the another file.
 * Why we should manage another `_highlightLayer` inside this class? Why we cannot use the original class for that?

Reply:
Point 1: Not sure about this one, both methods are actually different. Here's how it works basically:

- `RCTTextView` has 2 different layer ivars, one is a regular `CALayer` and the other one is a `CATiledLayer` subclass. The `configureLayer` method checks the current frame and creates an instance of the right type of layer based on it. It also removes the other layer type if an instance already exists (could happen when updating text). This means we always have one layer instanciated and the other one nil.
- `RCTTextRenderer` is a `CALayerDelegate` which handles all the text drawing, this delegate is used by both the `CALayer` and `CATiledLayer` this way there is no duplication of the text drawing code.

Point 2: Yea I didn't really know what it was initially and just ported all the code inside the `RCTTextRenderer` class but I agree it should still be managed by the `RCTTextView` class. I can probably make it a `CALayer` subclass `RCTHighlightLayer` to handle the highlight drawing logic separately.

Reply:
Point 1. This code draws `textStorage` on `context`, right? If so, why can't we decouple that into a static function?

Point 2. Not sure I understand why we need a subclass for that. Why can't we keep the code which renders highlight boxes where it is now?

Reply:
@shergin Oups I thought I replied to this -_-

1. Right now text drawing is abstracted in the `RCTTextRenderer` class. It implements `CALayerDelegate` so it can be set on any `CALayer`. The goal of this abstraction is to avoid duplicating the text drawing code since it is used by the 2 different `CALayer` implementations

In this case I don't see the need to abstract text drawing further in a static function since the drawing code is already implemented only once. An alternative would be to create 2 different `CALayer` subclasses and extract the drawing code in a static function like you mentioned instead of using the delegate.

The core of the code is the dance between the layer types in `configureLayers`, not sure if we want to come up with a cleaner abstraction but since I doubt there will be more than 2 layer types it should get much more complex.

2. Yea I didn't end up doing this, I moved it back to `RCTTextView`